### PR TITLE
HTML: Add support for declarative JSX elements

### LINKED ISSUE #11608: feat(html): add support for declarative JSX elements
Right now when exporting an element from a story we have to do this in an imperative way

```tsx
export const withEmoji = () => {
  const button = document.createElement('button');
  button.innerText = '😀 😎 👍 💯';
  return button;
};
```

Would be great if we could do this in an declarative way using JSX (events also work)

```tsx
export const withEmoji = () => {
  return <button>😀 😎 👍 💯</button>;
};
```

Current alternative for declarative exporting is to use string html, but in my opinion I think JSX exporting is better.

```tsx
export const withEmoji = () => {
  return `<button>😀 😎 👍 💯</button>`;
};
```

I can help implemet this feature

I worked on an [experimntal PR](https://github.com/storybookjs/storybook/pull/11072) in the past, that used React to render JSX to string.  
But now I think using [`dom-chef`](https://github.com/vadimdemedes/dom-chef) will be better, because it creates HTML elements instead of string template

### PR DESCRIPTION
**Pull request by unrevised6419**
Issue: #11608

Added support for declarative JSX elements

### CODE REVIEW

**Comment by iamandrewluca** [comment]
Should I document why we need `import React from 'dom-chef'` ?

**Comment by ndelangen** [comment]
Hi @iamandrewluca

Sorry for leaving this so untouched, we just passed a huge milestone: version 6.0.0, and this just wasn't the top priority.
Hopefully you're still interested, because this looks like a good addition to storybook for HTML to me.

Looks like next needs to be merged into this branch, and ensure tests pass. 2 merge conflicts are present, but I hope those are easy to fix. Would you be able to give that a go?

Hope to get this merged asap ❤️

**Comment by iamandrewluca** [comment]
@ndelangen this week I'll find some time to rebase on top of main branch and review again all changes to work well.

**Comment by iamandrewluca** [comment]
I rebased on top of next.

File: app/html/package.json [hunk 1]
```diff
@@ -35,8 +35,9 @@
   },
   "dependencies": {
     "@babel/plugin-transform-react-jsx": "^7.10.4",
-    "@storybook/addons": "6.0.12",
-    "@storybook/core": "6.0.12",
+    "@storybook/addons": "6.0.13",
+    "@storybook/core": "6.0.13",
+
```
Comment (shilman) [inline app/html/package.json:40]:
```suggestion
```

**Comment by NamHuynhHoang** [comment]
Các bạn nên thử xấp sếp lại toàn bộ tệp

**Comment by ndelangen** [comment]
> Các bạn nên thử xấp sếp lại toàn bộ tệp

You should try to re-stack the entire file

**Comment by iamandrewluca** [comment]
Do you want me to make a rebase on top of next branch?

**Comment by ndelangen** [comment]
No need, I'll take care of it

I all tests pass, I think we can go ahead and merge, if @shilman agrees.

**Comment by shilman** [comment]
@ndelangen This is a 6.1 feature, so I'd appreciate it if we can wait a few more days while we're stabilizing 6.0

**Comment by ndelangen** [comment]
@iamandrewluca can you take a look at the failing test:
```
FAIL  examples/html-kitchen-sink/tests/htmlshots.test.js
  ● Test suite failed to run

    storyshots found 0 stories

      106 |     }
      107 |     else {
    > 108 |         throw new Error('storyshots found 0 stories');
          |               ^
      109 |     }
      110 | }
      111 | exports.default = testStorySnapshots;

      at testStorySnapshots (addons/storyshots/storyshots-core/dist/api/index.js:108:15)
      at Object.<anonymous> (examples/html-kitchen-sink/tests/htmlshots.test.js:4:1)
```

**Comment by iamandrewluca** [comment]
@ndelangen at this moment I don't exactly know where to start. It seems storyshots need to be configured for jsx supoprt

File: app/html/src/server/framework-preset-html.ts [hunk 2]
```diff
@@ -20,3 +21,10 @@ export function webpack(config: Configuration) {
     },
   };
 }
+
+export function babelDefault(config: TransformOptions) {
+  return {
+    ...config,
+    plugins: [...config.plugins, require.resolve('@babel/plugin-transform-react-jsx')],
+  };
+}
```
Comment (ndelangen) [inline app/html/src/server/framework-preset-html.ts:30]: This needs to be applied to the babel config used to jest as well.

I've tried a few methods, but didn't get it to work..

Comment (iamandrewluca) [reply app/html/src/server/framework-preset-html.ts:30]: I'll take a look upcoming weekend. Thanks for suggestion 🚀

**Comment by iamandrewluca** [comment]
Rebased on top of `next`. Related to `storyshots found 0 stories`, I looked into other issues, and it seems more people encounter this issue.

**Pull request closed without merging**